### PR TITLE
[FEATURE] Docker container support

### PR DESCRIPTION
**Pull request by UXVirtual**
# Brief Summary
This PR adds support for running the Gradio interface for Voice-Clone-Studio inside a Docker container with [flash-attention](https://github.com/kingbri1/flash-attention). It also adds an "AI Denoise" button which will remove background noise from an uploaded clip in-place.

I tested this on Windows 11 with CUDA 13.0, but it should work with CUDA 12.8 too and may also be possible to run it on a Linux host, but I haven't tested that option. This PR also includes unit tests to cover usage the DeepFilterNet dependency and can validate if the runtime environment can perform denoising and is configured correctly.

It's advisable to run "AI Denoising" before hitting the "Normalize" and "Convert to Mono" button. This allows for clean clips to be used in zero-shot voice cloning when recordings have been made without expensive audio equipment.

Under the hood this uses [DeepFilterNet](https://github.com/Rikorose/DeepFilterNet) to remove background noise (repeating audio and background music). This is useful for improving the quality of the transcription and to avoid false-positives from spoken words in the background.

# Detailed Summary

This pull request introduces robust support for running the application in Docker, including GPU acceleration and compatibility with DeepFilterNet on modern PyTorch/Torchaudio versions. It adds a Dockerfile, Docker Compose configuration, and a compatibility patch for DeepFilterNet, along with integration tests and documentation updates. These changes make local and containerized deployment more reliable, especially for users running on Windows with WSL2 and NVIDIA GPUs.

**Dockerization and Deployment:**
- Added a `Dockerfile` using Python 3.12-slim, installing all necessary system and Python dependencies, including CUDA-specific PyTorch, Flash Attention, and Rust for DeepFilterNet. The build process applies compatibility patches and exposes the Gradio port for the application.
- Introduced `docker-compose.yaml` to orchestrate the container, map persistent volumes for samples, output, and HuggingFace models, and enable GPU access using NVIDIA runtime.
- Updated `.dockerignore` to exclude unnecessary files and directories from the build context, ensuring smaller and cleaner Docker images.

**DeepFilterNet Compatibility and Integration:**
- Added `patches/deepfilternet_torchaudio_patch.py`, a comprehensive monkey-patch to restore compatibility between DeepFilterNet and new Torchaudio versions by patching missing APIs and providing robust fallbacks for audio file handling.
- Updated `voice_clone_studio.py` to apply the patch at startup, gracefully handle DeepFilterNet import errors, and add lazy-loading logic for DeepFilterNet models and state. [[1]](diffhunk://#diff-bdb54e3fead8a79b3a0e5a00a9baf476815cc21d0271d493913f6fd8180e2d8bR23-R50) [[2]](diffhunk://#diff-bdb54e3fead8a79b3a0e5a00a9baf476815cc21d0271d493913f6fd8180e2d8bR77-R79) [[3]](diffhunk://#diff-bdb54e3fead8a79b3a0e5a00a9baf476815cc21d0271d493913f6fd8180e2d8bR193-R199) [[4]](diffhunk://#diff-bdb54e3fead8a79b3a0e5a00a9baf476815cc21d0271d493913f6fd8180e2d8bR229-R258)

**Testing and Documentation:**
- Added `tests/integration_test_denoiser.py`, a script to validate DeepFilterNet integration by generating dummy audio, running enhancement, and saving output, ensuring the patch and model work as expected.
- Expanded `README.md` with detailed instructions for Docker setup on Windows (including WSL2 and NVIDIA drivers), and documented how to run integration tests inside the container.

These changes collectively improve cross-platform reliability, make GPU-accelerated deployment straightforward, and ensure DeepFilterNet works out-of-the-box in both local and Docker environments.

### CODE REVIEW

**Comment by UXVirtual** [comment]
> Hi UXVirtual, any chance you could update your code to work with the latest version when you update it?

Yes, am planing to do that this weekend. Do you want me to rebase it on the dev branch?

**Comment by FranckyB** [comment]
Hi Joshua and Michael, I've just pushed to Dev Docker support using what you guys created.
I've also integrated DeepFilterNet as you had done Micheal.

I re-ordered the button to have Denoise first, as you recommended to do this as the first step.

I did a big push in the last 2 days, so quite a bit changed, added quite a bit of features that were asked.
Offline mode (With option to download models directly in app), removed Dependency to Flash Attention when Finetuning, New "Emotion System".  Advanced parameters for all.

The new emotion system is kind of a hack but can be intriguing, it basically tweaks the advanced parameters to get different results.  Added a way to save and delete new ones.  Night is still young, so I'll tweak a bit more 🤣

**Comment by FranckyB** [comment]
Pushed new update to Main.  🤞

**Comment by FranckyB** [comment]
Turns out working until 6:30 am is not ideal 🤣
2 small issues had creeped in.  So if you had pulled already, be sure to pull again.
A user mentioned issues with Normalize but I haven't been able to replicate.

**Comment by FranckyB** [comment]
Forgot to fix
server_name=os.getenv("GRADIO_SERVER_NAME", "127.0.0.1"),

now corrected and build properly.

**Comment by UXVirtual** [comment]
> Forgot to fix server_name=os.getenv("GRADIO_SERVER_NAME", "127.0.0.1"),
> 
> now corrected and build properly.

Thanks for that, I'm in the process of reviewing the updates now.

**Comment by FranckyB** [comment]
Ok, I'm seeing an issue with Finetuning.
My code is windows specific and worst yet, its hardcoded to the apps local venv.  
I imagine you guys had fixed that..
I'll take a look, not a big fix I imagine

**Comment by UXVirtual** [comment]
I discovered an issue with writing outputs during the Voice Design tab. I've made a PR for this fix here: https://github.com/FranckyB/Voice-Clone-Studio/pull/24

**Comment by UXVirtual** [comment]
Closing this PR in favor of the changes introduced in #24 and #25
